// @metaplex-foundation/mpl-toolbox/dist/src/generated/instructions/thawToken.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.thawToken = exports.getThawTokenInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getThawTokenInstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.u8)()]], {
        description: 'ThawTokenInstructionData',
    }), (value) => ({ ...value, discriminator: 11 }));
}
exports.getThawTokenInstructionDataSerializer = getThawTokenInstructionDataSerializer;
// Instruction.
function thawToken(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    // Resolved inputs.
    const resolvedAccounts = {
        account: [input.account, true],
        mint: [input.mint, false],
        owner: [input.owner, false],
    };
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.account, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.mint, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.owner, false);
    // Data.
    const data = getThawTokenInstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.thawToken = thawToken;
