// @metaplex-foundation/mpl-bubblegum/dist/src/generated/types/instructionName.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInstructionNameSerializer = exports.InstructionName = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var InstructionName;
(function (InstructionName) {
    InstructionName[InstructionName["Unknown"] = 0] = "Unknown";
    InstructionName[InstructionName["MintV1"] = 1] = "MintV1";
    InstructionName[InstructionName["Redeem"] = 2] = "Redeem";
    InstructionName[InstructionName["CancelRedeem"] = 3] = "CancelRedeem";
    InstructionName[InstructionName["Transfer"] = 4] = "Transfer";
    InstructionName[InstructionName["Delegate"] = 5] = "Delegate";
    InstructionName[InstructionName["DecompressV1"] = 6] = "DecompressV1";
    InstructionName[InstructionName["Compress"] = 7] = "Compress";
    InstructionName[InstructionName["Burn"] = 8] = "Burn";
    InstructionName[InstructionName["CreateTree"] = 9] = "CreateTree";
    InstructionName[InstructionName["VerifyCreator"] = 10] = "VerifyCreator";
    InstructionName[InstructionName["UnverifyCreator"] = 11] = "UnverifyCreator";
    InstructionName[InstructionName["VerifyCollection"] = 12] = "VerifyCollection";
    InstructionName[InstructionName["UnverifyCollection"] = 13] = "UnverifyCollection";
    InstructionName[InstructionName["SetAndVerifyCollection"] = 14] = "SetAndVerifyCollection";
    InstructionName[InstructionName["MintToCollectionV1"] = 15] = "MintToCollectionV1";
    InstructionName[InstructionName["SetDecompressableState"] = 16] = "SetDecompressableState";
})(InstructionName = exports.InstructionName || (exports.InstructionName = {}));
function getInstructionNameSerializer(_context = {}) {
    return (0, serializers_1.scalarEnum)(InstructionName, {
        description: 'InstructionName',
    });
}
exports.getInstructionNameSerializer = getInstructionNameSerializer;
