// @metaplex-foundation/mpl-bubblegum/dist/src/generated/instructions/unverifyCreator.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.unverifyCreator = exports.getUnverifyCreatorInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUnverifyCreatorInstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 8 })],
        ['root', (0, serializers_1.bytes)({ size: 32 })],
        ['dataHash', (0, serializers_1.bytes)({ size: 32 })],
        ['creatorHash', (0, serializers_1.bytes)({ size: 32 })],
        ['nonce', (0, serializers_1.u64)()],
        ['index', (0, serializers_1.u32)()],
        ['metadata', (0, types_1.getMetadataArgsSerializer)()],
    ], { description: 'UnverifyCreatorInstructionData' }), (value) => ({
        ...value,
        discriminator: [107, 178, 57, 39, 105, 115, 112, 152],
    }));
}
exports.getUnverifyCreatorInstructionDataSerializer = getUnverifyCreatorInstructionDataSerializer;
// Instruction.
function unverifyCreator(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplBubblegum', 'BGUMAp9Gq7iTEuizy4pqaxsTyUCBK68MDfK752saRPUY');
    // Resolved inputs.
    const resolvedAccounts = {
        leafOwner: [input.leafOwner, false],
        merkleTree: [input.merkleTree, true],
        creator: [input.creator, false],
    };
    const resolvingArgs = {};
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'treeConfig', input.treeConfig
        ? [input.treeConfig, false]
        : [
            (0, accounts_1.findTreeConfigPda)(context, {
                merkleTree: (0, umi_1.publicKey)(input.merkleTree, false),
            }),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'leafDelegate', input.leafDelegate
        ? [input.leafDelegate, false]
        : [input.leafOwner, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'payer', input.payer
        ? [input.payer, false]
        : [context.payer, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'logWrapper', input.logWrapper
        ? [input.logWrapper, false]
        : [
            context.programs.getPublicKey('splNoop', 'noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'compressionProgram', input.compressionProgram
        ? [input.compressionProgram, false]
        : [
            context.programs.getPublicKey('splAccountCompression', 'cmtDvXumGCrqC1Age74AVPhSRVXJMd8PJS91L8KbNCK'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'systemProgram', input.systemProgram
        ? [input.systemProgram, false]
        : [
            context.programs.getPublicKey('splSystem', '11111111111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvingArgs, 'dataHash', input.dataHash ??
        (0, hooked_1.resolveDataHash)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, false));
    (0, shared_1.addObjectProperty)(resolvingArgs, 'creatorHash', input.creatorHash ??
        (0, hooked_1.resolveCreatorHash)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, false));
    (0, shared_1.addObjectProperty)(resolvingArgs, 'proof', input.proof ?? []);
    const resolvedArgs = { ...input, ...resolvingArgs };
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.treeConfig, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.leafOwner, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.leafDelegate, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.merkleTree, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.payer, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.creator, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.logWrapper, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.compressionProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.systemProgram, false);
    // Remaining Accounts.
    const remainingAccounts = resolvedArgs.proof.map((address) => [address, false]);
    remainingAccounts.forEach((remainingAccount) => (0, shared_1.addAccountMeta)(keys, signers, remainingAccount, false));
    // Data.
    const data = getUnverifyCreatorInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.unverifyCreator = unverifyCreator;
