// @metaplex-foundation/mpl-token-metadata/dist/src/generated/instructions/revokeCollectionAuthority.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.revokeCollectionAuthority = exports.getRevokeCollectionAuthorityInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getRevokeCollectionAuthorityInstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.u8)()]], { description: 'RevokeCollectionAuthorityInstructionData' }), (value) => ({ ...value, discriminator: 24 }));
}
exports.getRevokeCollectionAuthorityInstructionDataSerializer = getRevokeCollectionAuthorityInstructionDataSerializer;
// Instruction.
function revokeCollectionAuthority(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
    // Resolved inputs.
    const resolvedAccounts = {
        collectionAuthorityRecord: [input.collectionAuthorityRecord, true],
        delegateAuthority: [input.delegateAuthority, true],
        revokeAuthority: [input.revokeAuthority, true],
        mint: [input.mint, false],
    };
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'metadata', input.metadata
        ? [input.metadata, false]
        : [
            (0, accounts_1.findMetadataPda)(context, { mint: (0, umi_1.publicKey)(input.mint, false) }),
            false,
        ]);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionAuthorityRecord, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.delegateAuthority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.revokeAuthority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.metadata, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.mint, false);
    // Data.
    const data = getRevokeCollectionAuthorityInstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.revokeCollectionAuthority = revokeCollectionAuthority;
