// @metaplex-foundation/mpl-token-metadata/dist/src/generated/instructions/approveCollectionAuthority.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.approveCollectionAuthority = exports.getApproveCollectionAuthorityInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getApproveCollectionAuthorityInstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.u8)()]], { description: 'ApproveCollectionAuthorityInstructionData' }), (value) => ({ ...value, discriminator: 23 }));
}
exports.getApproveCollectionAuthorityInstructionDataSerializer = getApproveCollectionAuthorityInstructionDataSerializer;
// Instruction.
function approveCollectionAuthority(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
    // Resolved inputs.
    const resolvedAccounts = {
        collectionAuthorityRecord: [input.collectionAuthorityRecord, true],
        newCollectionAuthority: [input.newCollectionAuthority, false],
        mint: [input.mint, false],
        rent: [input.rent, false],
    };
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'updateAuthority', input.updateAuthority
        ? [input.updateAuthority, true]
        : [context.identity, true]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'payer', input.payer
        ? [input.payer, true]
        : [context.payer, true]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'metadata', input.metadata
        ? [input.metadata, false]
        : [
            (0, accounts_1.findMetadataPda)(context, { mint: (0, umi_1.publicKey)(input.mint, false) }),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'systemProgram', input.systemProgram
        ? [input.systemProgram, false]
        : [
            context.programs.getPublicKey('splSystem', '11111111111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionAuthorityRecord, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.newCollectionAuthority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.updateAuthority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.payer, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.metadata, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.mint, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.systemProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.rent, true);
    // Data.
    const data = getApproveCollectionAuthorityInstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.approveCollectionAuthority = approveCollectionAuthority;
