// @metaplex-foundation/mpl-token-metadata/dist/src/generated/instructions/bubblegumSetCollectionSize.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.bubblegumSetCollectionSize = exports.getBubblegumSetCollectionSizeInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getBubblegumSetCollectionSizeInstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['setCollectionSizeArgs', (0, types_1.getSetCollectionSizeArgsSerializer)()],
    ], { description: 'BubblegumSetCollectionSizeInstructionData' }), (value) => ({ ...value, discriminator: 36 }));
}
exports.getBubblegumSetCollectionSizeInstructionDataSerializer = getBubblegumSetCollectionSizeInstructionDataSerializer;
// Instruction.
function bubblegumSetCollectionSize(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
    // Resolved inputs.
    const resolvedAccounts = {
        collectionMetadata: [input.collectionMetadata, true],
        collectionAuthority: [input.collectionAuthority, true],
        collectionMint: [input.collectionMint, false],
        bubblegumSigner: [input.bubblegumSigner, false],
        collectionAuthorityRecord: [
            input.collectionAuthorityRecord,
            false,
        ],
    };
    const resolvingArgs = {};
    const resolvedArgs = { ...input, ...resolvingArgs };
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionMetadata, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionAuthority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionMint, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.bubblegumSigner, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.collectionAuthorityRecord, true);
    // Data.
    const data = getBubblegumSetCollectionSizeInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.bubblegumSetCollectionSize = bubblegumSetCollectionSize;
