// @metaplex-foundation/mpl-token-metadata/dist/src/generated/instructions/transferV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferV1 = exports.getTransferV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getTransferV1InstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['transferV1Discriminator', (0, serializers_1.u8)()],
        ['amount', (0, serializers_1.u64)()],
        ['authorizationData', (0, serializers_1.option)((0, types_1.getAuthorizationDataSerializer)())],
    ], { description: 'TransferV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 49,
        transferV1Discriminator: 0,
        amount: value.amount ?? 1,
        authorizationData: value.authorizationData ?? (0, umi_1.none)(),
    }));
}
exports.getTransferV1InstructionDataSerializer = getTransferV1InstructionDataSerializer;
// Instruction.
function transferV1(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
    // Resolved inputs.
    const resolvedAccounts = {
        destinationOwner: [input.destinationOwner, false],
        mint: [input.mint, false],
    };
    const resolvingArgs = {};
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'tokenOwner', input.tokenOwner
        ? [input.tokenOwner, false]
        : [context.identity.publicKey, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'token', input.token
        ? [input.token, true]
        : [
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, umi_1.publicKey)(input.mint, false),
                owner: (0, umi_1.publicKey)(resolvedAccounts.tokenOwner[0], false),
            }),
            true,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'destinationToken', input.destinationToken
        ? [input.destinationToken, true]
        : [
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, umi_1.publicKey)(input.mint, false),
                owner: (0, umi_1.publicKey)(input.destinationOwner, false),
            }),
            true,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'metadata', input.metadata
        ? [input.metadata, true]
        : [
            (0, accounts_1.findMetadataPda)(context, { mint: (0, umi_1.publicKey)(input.mint, false) }),
            true,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'edition', input.edition
        ? [input.edition, false]
        : (0, hooked_1.resolveMasterEditionForProgrammables)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, false));
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'tokenRecord', input.tokenRecord
        ? [input.tokenRecord, true]
        : (0, hooked_1.resolveTokenRecord)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, true));
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'destinationTokenRecord', input.destinationTokenRecord
        ? [input.destinationTokenRecord, true]
        : (0, hooked_1.resolveDestinationTokenRecord)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, true));
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authority', input.authority
        ? [input.authority, false]
        : [context.identity, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'payer', input.payer
        ? [input.payer, true]
        : [context.payer, true]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'systemProgram', input.systemProgram
        ? [input.systemProgram, false]
        : [
            context.programs.getPublicKey('splSystem', '11111111111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'sysvarInstructions', input.sysvarInstructions
        ? [input.sysvarInstructions, false]
        : [
            (0, umi_1.publicKey)('Sysvar1nstructions1111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'splTokenProgram', input.splTokenProgram
        ? [input.splTokenProgram, false]
        : [
            context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'splAtaProgram', input.splAtaProgram
        ? [input.splAtaProgram, false]
        : [
            context.programs.getPublicKey('splAssociatedToken', 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authorizationRules', input.authorizationRules
        ? [input.authorizationRules, false]
        : [programId, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authorizationRulesProgram', input.authorizationRulesProgram
        ? [input.authorizationRulesProgram, false]
        : (0, hooked_1.resolveAuthorizationRulesProgram)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, false));
    const resolvedArgs = { ...input, ...resolvingArgs };
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.token, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.tokenOwner, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.destinationToken, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.destinationOwner, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.mint, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.metadata, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.edition, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.tokenRecord, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.destinationTokenRecord, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.payer, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.systemProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.sysvarInstructions, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.splTokenProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.splAtaProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authorizationRulesProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authorizationRules, false);
    // Data.
    const data = getTransferV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.transferV1 = transferV1;
