// @metaplex-foundation/mpl-token-metadata/dist/src/generated/instructions/updateAsProgrammableConfigDelegateV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAsProgrammableConfigDelegateV2 = exports.getUpdateAsProgrammableConfigDelegateV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUpdateAsProgrammableConfigDelegateV2InstructionDataSerializer(_context = {}) {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['updateAsProgrammableConfigDelegateV2Discriminator', (0, serializers_1.u8)()],
        ['ruleSet', (0, types_1.getRuleSetToggleSerializer)()],
        ['authorizationData', (0, serializers_1.option)((0, types_1.getAuthorizationDataSerializer)())],
    ], { description: 'UpdateAsProgrammableConfigDelegateV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 50,
        updateAsProgrammableConfigDelegateV2Discriminator: 5,
        ruleSet: value.ruleSet ?? (0, types_1.ruleSetToggle)('None'),
        authorizationData: value.authorizationData ?? (0, umi_1.none)(),
    }));
}
exports.getUpdateAsProgrammableConfigDelegateV2InstructionDataSerializer = getUpdateAsProgrammableConfigDelegateV2InstructionDataSerializer;
// Instruction.
function updateAsProgrammableConfigDelegateV2(context, input) {
    const signers = [];
    const keys = [];
    // Program ID.
    const programId = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
    // Resolved inputs.
    const resolvedAccounts = {
        token: [input.token, false],
        mint: [input.mint, false],
    };
    const resolvingArgs = {};
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authority', input.authority
        ? [input.authority, false]
        : [context.identity, false]);
    (0, shared_1.addObjectProperty)(resolvingArgs, 'delegateMint', input.delegateMint ?? (0, umi_1.publicKey)(input.mint, false));
    (0, shared_1.addObjectProperty)(resolvingArgs, 'delegateUpdateAuthority', input.delegateUpdateAuthority ?? context.identity.publicKey);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'delegateRecord', input.delegateRecord
        ? [input.delegateRecord, false]
        : [
            (0, accounts_1.findMetadataDelegateRecordPda)(context, {
                mint: resolvingArgs.delegateMint,
                delegateRole: types_1.MetadataDelegateRole.ProgrammableConfig,
                updateAuthority: resolvingArgs.delegateUpdateAuthority,
                delegate: (0, umi_1.publicKey)(resolvedAccounts.authority[0], false),
            }),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'metadata', input.metadata
        ? [input.metadata, true]
        : [
            (0, accounts_1.findMetadataPda)(context, { mint: (0, umi_1.publicKey)(input.mint, false) }),
            true,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'edition', input.edition
        ? [input.edition, false]
        : [programId, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'payer', input.payer
        ? [input.payer, true]
        : [context.payer, true]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'systemProgram', input.systemProgram
        ? [input.systemProgram, false]
        : [
            context.programs.getPublicKey('splSystem', '11111111111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'sysvarInstructions', input.sysvarInstructions
        ? [input.sysvarInstructions, false]
        : [
            (0, umi_1.publicKey)('Sysvar1nstructions1111111111111111111111111'),
            false,
        ]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authorizationRules', input.authorizationRules
        ? [input.authorizationRules, false]
        : [programId, false]);
    (0, shared_1.addObjectProperty)(resolvedAccounts, 'authorizationRulesProgram', input.authorizationRulesProgram
        ? [input.authorizationRulesProgram, false]
        : (0, hooked_1.resolveAuthorizationRulesProgram)(context, { ...input, ...resolvedAccounts }, { ...input, ...resolvingArgs }, programId, false));
    const resolvedArgs = { ...input, ...resolvingArgs };
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authority, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.delegateRecord, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.token, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.mint, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.metadata, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.edition, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.payer, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.systemProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.sysvarInstructions, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authorizationRulesProgram, false);
    (0, shared_1.addAccountMeta)(keys, signers, resolvedAccounts.authorizationRules, false);
    // Data.
    const data = getUpdateAsProgrammableConfigDelegateV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.updateAsProgrammableConfigDelegateV2 = updateAsProgrammableConfigDelegateV2;
