// @metaplex-foundation/mpl-token-metadata/dist/src/generated/shared/index.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAccountMeta = exports.addObjectProperty = void 0;
const umi_1 = require("@metaplex-foundation/umi");
/**
 * Helper function that dynamically updates the type of
 * an object as we add more properties to the object.
 * @internal
 */
function addObjectProperty(obj, key, value) {
    obj[key] = value;
}
exports.addObjectProperty = addObjectProperty;
function addAccountMeta(keys, signers, account, isOptional) {
    if (isOptional && !account[0]) {
        return;
    }
    if (!account[0]) {
        throw new Error('Expected instruction account to be defined');
    }
    if ((0, umi_1.isSigner)(account[0])) {
        signers.push(account[0]);
    }
    keys.push({
        pubkey: (0, umi_1.publicKey)(account[0], false),
        isSigner: (0, umi_1.isSigner)(account[0]),
        isWritable: account[1],
    });
}
exports.addAccountMeta = addAccountMeta;
